align_system/algorithms/alignment_adm_component.py

```diff
@@ -47,3 +49,94 @@ def run(self,
+        self.attributes = attributes
+
+    def run_returns(self):
+        return ('chosen_choice',
+                'best_sample_idx')
+
+    def run(
+        self,
+        attribute_prediction_scores,
+        alignment_target,
+    ):
+        """ Align based on medical urgency/KDMA tradeoff """
+        med_urg_str  = "Medical Urgency"
+        if alignment_target is None:
+            raise RuntimeError(
+                "Assumption violated: `alignment_target` was None"
+            )
+
+        target_kdmas = alignment_target_to_attribute_targets(
+            alignment_target,
+            self.attributes)
+
+        target_kdmas = [dict(t) for t in target_kdmas]
+        if len(target_kdmas) > 1:
+            raise RuntimeError("Multi-kdma alignment has not yet been implemented")
+
+        def _handle_single_value(predictions):
+            if not isinstance(predictions, list):
+                return [predictions]
+            return predictions
+
+        choices = list(attribute_prediction_scores.keys())
+        if len(choices) == 1:
+            return choices[0]
+        elif len(choices) > 2:
+            raise RuntimeError("This alignment function has not yet been implemented for >2 choices")
+
+        # Compute averages of predicted values
+        predictions = []
+        for choice, all_kdma_predictions in attribute_prediction_scores.items():
+            pred_dict = { "choice": choice }
+
+            # Get medical urgency
+            if med_urg_str not in all_kdma_predictions:
+                raise RuntimeError("Medical Urgency predictions required for this alignment function")
+            medical_urgency_preds = _handle_single_value(all_kdma_predictions[med_urg_str])
+            pred_dict[med_urg_str] = sum(medical_urgency_preds) / len(medical_urgency_preds)
+
+            # Get KDMA predictions relevant to target
+            kdma_preds = {}
+            for target_kdma in target_kdmas:
+                kdma = target_kdma["kdma"]
+                kdma_preds = _handle_single_value(all_kdma_predictions[kdma])
+                kdma_preds[kdma] = sum(kdma_preds) / len(kdma_preds)
+            pred_dict["kdmas"] = kdma_preds
+
+            predictions.append(pred_dict)
+
+        # Sort by medical urgency
+        predictions.sort(key=lambda pred: pred[med_urg_str], reverse=True)
+
+        # Guaranteed to only have 2 choices at this point due to earlier checks
+        opt_a, opt_b = predictions[0], predictions[1]
+        medical_delta = opt_a[med_urg_str] - opt_b[med_urg_str]
+
+        # Guaranteed to only have 1 attribute due to earlier checks
+        attr = target_kdmas[0]["kdma"]
+        attribute_delta = opt_a["kdmas"][attr] - opt_b["kdmas"][attr]
+        if attribute_delta > 0:
+            # opt_a is more medically urgent AND more deserving based on attribute
+            return opt_a["choice"]
+
+        # Equation from ADEPT
+        probe_midpoint = 0.5 + (medical_delta - attribute_delta)/2
```

Comment:
I would want to know in the logs what the `probe_midpoint` is (maybe with `log.explain` with some text around it; could even say something about how the target is above or below the midpoint)